File: typescript/packages/core/src/evals/stub-queue.ts
import { z } from "zod";
import { type RecordedStubs, recordedStubs } from "../loop";

// The stub entries a run answers from (spec lane 32, A.3). stubs.json numbers occurrences per
// (tool, args_hash) across the whole recorded conversation, prefix entries first and marked
// `scope: "prefix"`. Picking a scope renumbers what is left from 0, so each key stays a queue
// the run consumes in log order.

const Entry = z.strictObject({
  tool: z.string(),
  args_hash: z.string(),
  occurrence: z.int().min(0),
  output: z.string(),
  is_error: z.boolean().optional(),
  scope: z.literal("prefix").optional(),
});
const StubScript = z.strictObject({ stubs: z.array(Entry) });

/** turn: the saved turn's entries only (offline, and a continued prefix). */
export type StubScope = "turn" | "conversation";

export type StubQueue = RecordedStubs & {
  /** The tool of the first call no entry answered, for `unmatched_external_op`. */
  readonly firstUnmatched: () => string | undefined;
};

export function stubQueue(script: unknown, scope: StubScope): StubQueue {
  const { stubs } = StubScript.parse(script);
  const picked =
    scope === "conversation"
      ? stubs
      : stubs.filter((s) => s.scope === undefined);
  const seen = new Map<string, number>();
  const queue = picked.map(({ tool, args_hash, output, is_error }) => {
    const key = `${tool}\n${args_hash}`;
    const occurrence = seen.get(key) ?? 0;
    seen.set(key, occurrence + 1);
    return {
      tool,
      args_hash,
      occurrence,
      output,
      ...(is_error === undefined ? {} : { is_error }),
    };
  });
  const inner = recordedStubs({ stubs: queue });
  let missed: string | undefined;
  return {
    ...inner,
    answer: (tool, argsHash) => {
      const got = inner.answer(tool, argsHash);
      if (got === undefined && missed === undefined) missed = tool;
      return got;
    },
    firstUnmatched: () => missed,
  };
}

File: python/src/threads/evals/stub_queue.py
"""The stub entries a run answers from (spec lane 32, A.3). `stubs.json` numbers occurrences per
(tool, args_hash) across the whole recorded conversation, prefix entries first and marked
`scope: "prefix"`. Picking a scope renumbers what is left from 0, so each key stays a queue the
run consumes in log order."""

from collections.abc import Mapping
from typing import Literal

from pydantic import JsonValue

from threads.loop.stubs import Stub, StubGateway, parse_stubs

type StubScope = Literal["turn", "conversation"]
"""turn: the saved turn's entries only (offline, and a continued prefix)."""


def _scoped(script: Mapping[str, JsonValue], scope: StubScope) -> list[JsonValue]:
    stubs = script.get("stubs")
    if not isinstance(stubs, list):
        raise ValueError("a stub script needs a stubs list")
    kept: list[JsonValue] = []
    for entry in stubs:
        if not isinstance(entry, dict):
            raise ValueError("a stub is an object")
        if scope == "turn" and entry.get("scope") is not None:
            continue
        kept.append({k: v for k, v in entry.items() if k != "scope"})
    return kept


def stub_queue(script: Mapping[str, JsonValue], scope: StubScope) -> tuple[Stub, ...]:
    """The entries of `scope`, renumbered from 0 per (tool, args_hash) in their file order."""
    seen: dict[tuple[str, str], int] = {}
    parsed = parse_stubs({"stubs": _scoped(script, scope)})
    out: list[Stub] = []
    for stub in parsed:
        key = (stub.tool, stub.args_hash)
        occurrence = seen.get(key, 0)
        seen[key] = occurrence + 1
        out.append(Stub(stub.tool, stub.args_hash, occurrence, stub.output, stub.is_error))
    return tuple(out)


def stub_gateway(script: Mapping[str, JsonValue], scope: StubScope) -> StubGateway:
    """One gateway for a whole simulated conversation, so its queue spans every turn."""
    return StubGateway(stub_queue(script, scope))
